docs/proposals/role_decomposition.md

```diff
@@ -0,0 +1,322 @@
+# Scaffolding for decomposing large roles
+
+## Why?
+
+Currently we have roles that are very large and encompass a lot of different
+components. This makes for a lot of logic required within the role, can
+create complex conditionals, and increases the learning curve for the role.
+
+## How?
+
+Creating a guide on how to approach breaking up a large role into smaller,
+component based, roles. Also describe how to develop new roles, to avoid creating
+large roles.
+
+## Proposal
+
+Create a new guide or append to the current contributing guide a process for
+identifying large roles that can be split up, and how to compose smaller roles
+going forward.
+
+### Large roles
+
+A role should be considered for decomposition if it:
+
+1) Configures/installs more than one product.
+1) Can configure multiple variations of the same product that can live
+side by side.
+1) Has different entry points for upgrading and installing a product
+
+Large roles should be responsible for:
+
+1) Composing smaller roles to provide a full solution such as an Openshift Master
```

Comment:
Again, I really like the idea. I am just sceptic wrt. how we are going to decompose the current roles, not how we are going to compose the current roles (possible new) into large roles. I have noticed some pattern in naming of roles:
- `etcd ` && `openshift_etcd`
- `etcd_client_certificates` && `openshift_etcd_client_certificates`
- `etcd_server_certificates` && `openshift_etcd_server_certificates`
- ...

A role with `openshift_` prefix just adds `openshift_facts` into the `meta/main.yml`, nothing else.
Depending on when the role is used it can have multiple wrappers, each with a list of additional dependencies.